src/core/eme/create_session.ts

```diff
@@ -0,0 +1,217 @@
+    keySystemAccess,
+    mediaKeys,
+    sessionStorage,
+  } = mediaKeysInfos;
+
+  if (mediaKeys.createSession == null) {
+    throw new Error("Invalid MediaKeys implementation: Missing createSession");
+  }
+
+  const mksConfig = keySystemAccess.getConfiguration();
+  const sessionTypes = mksConfig.sessionTypes;
+  const hasPersistence = (
+    sessionTypes && arrayIncludes(sessionTypes, "persistent-license")
+  );
+
+  // XXX TODO Should we create a "persistent-license" session if we have no
+  // sessionStorage as we have no means to load the session in that case?
+  const sessionType = hasPersistence && keySystemOptions.persistentLicense ?
+    "persistent-license" : "temporary";
+
+  log.debug(`eme: create a new ${sessionType} session`);
+
+  const session = (mediaKeys as any /* TS bug */).createSession(sessionType);
+  $loadedSessions.add(initData, initDataType, session);
+
+  if (!hasPersistence || !sessionStorage || !keySystemOptions.persistentLicense) {
+    return Observable.of({
+      type: "created-session" as "created-session",
+      value: { mediaKeySession: session },
+    });
+  }
+
+  const storedEntry = sessionStorage.get(initData, initDataType);
+  if (!storedEntry) {
+    return Observable.of({
+      type: "created-session" as "created-session",
+      value: { mediaKeySession: session },
+    });
+  }
+
+  return loadPersistentSession(storedEntry.sessionId, session)
+    .map((hasLoadedSession) => {
+      if (!hasLoadedSession) {
+        log.warn("eme: no data stored for the loaded session");
+        sessionStorage.delete(initData, initDataType);
+        return {
+          type: "created-session" as "created-session",
+          value: { mediaKeySession: session },
+        };
+      }
+
+      if (hasLoadedSession && isSessionUsable(session)) {
+        sessionStorage.add(initData, initDataType, session);
+        return {
+          type: "loaded-persistent-session" as "loaded-persistent-session",
+          value: { mediaKeySession: session },
+        };
+      }
+
+      // Unusable persistent session: recreate a new session from scratch.
+      $loadedSessions.closeSession(session);
+      sessionStorage.delete(initData, initDataType);
+
+      const newSession =
+        (mediaKeys as any /* TS bug */).createSession("persistent-license");
+      $loadedSessions.add(initData, initDataType, newSession);
+
+      return {
+        type: "created-session" as "created-session",
+        value: { mediaKeySession: newSession },
+      };
+    })
+    .catch(() => {
+      // Failure to load persistent session: recreate a new session from scratch.
+      $loadedSessions.closeSession(session);
+      sessionStorage.delete(initData, initDataType);
+
+      const newSession =
+        (mediaKeys as any /* TS bug */).createSession("persistent-license");
+      $loadedSessions.add(initData, initDataType, newSession);
+
+      return Observable.of({
+        type: "created-session" as "created-session",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=187&issues=AWMrRtWamtQEh111nXqA&open=AWMrRtWamtQEh111nXqA)

[ISSUE_KEY]: AWMrRtWamtQEh111nXqA